### PR TITLE
plugins for drivers

### LINKED ISSUE #432: Support custom drivers to connect to data APIs
### Kind of request

Adding new functionality

### Enhancement Description

We want to enable users to access data directly from custom APIs through the data catalog.  
This is especially relevant for observational data that gets updated frequently, but can be used in a broader context.

### Use case

There are two known use cases in plugins

- HydroMT-Wflow does uses the global water watch api: https://github.com/Deltares/hydromt_wflow/blob/74812afe1ad163258c1afe63bfecdb6a14e7900d/hydromt_wflow/workflows/waterbodies.py#L256
- HydroMT-FIAT uses an api for NSI data: https://github.com/Deltares/hydromt_fiat/blob/fiat_integrator/hydromt_fiat/workflows/exposure_vector.py#L125

### Additional Context

My initial thoughts:
-	Downloading data can be supported through custom “driver” functions which can be made available to the data adaptors via the entrypoints (similar to plugins). We need to provide a template for these functions. The data catalog file can be used to pass possible kwargs to these driver functions similar to the current drivers.
-	The actual functions should live in plugins and/or a separate package to avoid having to maintain these and related dependencies in HydroMT.
-	I think it would be good practice to cache the downloaded data in a subfolder of the model root with a yml file for reproducibility.

### PR DESCRIPTION
**Pull request by Jaapel**
## Issue addressed
Fixes #432 

## Explanation
Added plugin behaviour for drivers. To load drivers, you only need to import them, as the validators search for subclasses of the Driver Component. To keep the plugin consistent with others, I did keep the `PLUGINS` API similar to the other plugins, although it is not used in the `BaseDriver` class.

The main advantage of the current `BaseDriver` class is that you could define a `Driver` inline, and still use `BaseDriver.model_validate`, or `create_source({"driver": {"name": "my_plugin_driver"}})` without having to do an editable install.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `v1`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.rst if needed

### CODE REVIEW

File: hydromt/plugins.py
Comment (deltamarnix) [inline hydromt/plugins.py:95]: https://docs.python.org/3/library/inspect.html#inspect.isabstract

Comment (deltamarnix) [inline hydromt/plugins.py:95]:
```suggestion
        return {name: value["type"] for name, value in self._driver_plugins.items() if not inspect.isabstract(value["type"])}
```

File: hydromt/drivers/__init__.py
Comment (deltamarnix) [inline hydromt/drivers/__init__.py:1]: I see that only the drivers are in this list of `__all__`, but there are some other python files that are now not exposed. When someone wants to create its own driver, you might want to open up preprocessor.py.

And also `BaseDriver` is now part of the list of drivers, although it is an abstract class.

Comment (Jaapel) [reply hydromt/drivers/__init__.py:1]: How do we see `__all__`? Just for the entrypoints, or just as a way to shorten import paths?

**Comment by Jaapel** [comment]
@deltamarnix @savente93 Added: https://github.com/Deltares/hydromt/issues/872

File: hydromt/_typing/type_def.py
Comment (savente93) [inline hydromt/_typing/type_def.py:59]: if it's no longer necessary I'd remove it.

File: hydromt/drivers/preprocessing.py
Comment (savente93) [inline hydromt/drivers/preprocessing.py:1]: Sonar cube is flagging that you coppied some code here from data dapter without removing it from the original place. Can you remove it in either place? Same for Pygrio_driver and gis_utils

File: hydromt/metadata_resolver/convention_resolver.py
Comment (savente93) [inline hydromt/metadata_resolver/convention_resolver.py:77]: What about this TODO? Can we remove it or either link an issue?

**Review by savente93 — COMMENTED** [review]
Couple of small comments, but otherwise LGTM, once you solve those I'll be happy to approve. Tests look good so if they pass I think it's all good enough to go

File: hydromt/drivers/preprocessing.py
Comment (Jaapel) [reply hydromt/drivers/preprocessing.py:1]: It is currently present in the unmigrated version of the `DataAdapter`. I can remove it if you want, but that makes the unmigrated code harder to read.

Comment (savente93) [reply hydromt/drivers/preprocessing.py:1]: Then it can stay, we'll just remove it when we migrate that part of the code.

**Review by savente93 — APPROVED** [review]
LGTM!

**Pull request merged**
